config/config.exs

```diff
@@ -32,6 +32,7 @@ use Mix.Config
 config :crawly, Crawly.Worker, client: HTTPoison
 
 config :crawly,
+  max_retries: 3,
```

Comment:
Can we discuss the problem with global configs again, please? I am kind of missing the key problem there. 

E.g. do we have practical concerns when several applications would use different versions of Crawly as a dependency?

Reply:
The issues with global configs that I can think of are:
1. a single application can be scraping multiple sites for different information, each might have retry requirements or different user agents required, so it makes sense to pass options directly to each pipeline such that the configs don't interfere with another pipeline.
2. Reduces reusability. For example, if I were to create a custom pipeline module that can apply a set of pipelines to an item if it meets a certain condition, for example, i can't reuse the `Pipelines.Validate` module because takes its config from the global config. In the case of retries, say i only want to apply retries to a particular type of scraped item, or if the url matches a certain condition, I can't simply create a pipeline that does a conditional if that then applies the pipeline.
For example, with direct options being passed, I can do this quite easily
```elixir
def MyPipelineExtension do
    def run( %{nested_item: nested } = item, state, opts) do
        Crawly.Utils.pipe([
            {Crawly.Pipeline.Validate, fields: [:my, :required, :fields] }
        ], item, state)
    end
    def run(item, state, _opts), do: {item, state}
end
```
It makes the built in modules so much more useful, and they act more like building blocks that can be applied according to the developer's needs. Using a global config is more like saying that the building block can only be done in one particular way, but the developer may actually need to adapt it for a whole different purpose

That's why i don't like global configs, because they make the built in modules essentially useless to me because I can't extend them like add special conditions or side effects.

Reply:
The example is a simplification, of course, but the idea is that it allows more flexibility in declarations. So for example, if in the future we implement spider-level pipeline/middleware overrides, it would be just a simple pipeline that is customized for the site, instead of having to put_env and get_env and all that, which would change every pipeline's behaviour in the application.

Reply:
> a single application can be scraping multiple sites for different information, each might have retry requirements or different user agents required, so it makes sense to pass options directly to each pipeline such that the configs don't interfere with another pipeline.

That part is clear. I am thinking of adding the per-spider configs ASAP. (We had it in Scrapy, and I will add it here as well!)

> Reduces reusability. For example, if I were to create a custom pipeline module that can apply a set of pipelines to an item if it meets a certain condition, for example, i can't reuse the Pipelines. Validate module because takes its config from the global config. In the case of retries, say i only want to apply retries to a particular type of scraped item, or if the url matches a certain condition, I can't simply create a pipeline that does a conditional if that then applies the pipeline.
For example, with direct options being passed, I can do this quite easily ....

Yes, at some extent it's the case. But I would not expect Crawly to be used as a dependency somewhere, where some other app has crawly as a dependency as well.
Let's maybe keep some balance here. I agree that with middlewares/pipelines it's quite nice, as it makes the config cleaner.

Reply:
Yup of course, i understand what you mean about clean configs. But perhaps an escape hatch would be good for more advanced use cases, which is where support for the tuple-based would be good (so that new users can use the global configs, and when they need can switch over)